statistics_epic_1_of_3

Depechist opened the pull request:
- Выбранная командой архитектура: MVVM
- Верстка: кодом
- Декомпозиция эпика (перешли из Яндекс Трекера в Notion): https://vadamask.notion.site/88b63082858c472ba96a17224134b9b9?v=0f49dd80115a48b2899b792c70b54117&pvs=4
- Также декомпозиция всей команды в файле Docs\Decomposition Проекта.
- Первоначальная оценка задач в указанных выше локациях.
- Реальное время, потраченное на выполнение 1/3: к сожалению, так и не поняли как запускать/ставить на паузу задачу из работы на доске Яндекс.Трекере, которым пользовались до Notionm поэтому указываю также приблизительное затраченное время ниже (постоянно переключаюсь между работой и дипломом).

СТАТИСТИКА, декомпозиция (по экранам)

- Экран Рейтинга (7 ч) - итого около 12 часов фактически
    - Верстка (4 ч) - около 10 часов фактически с сетью
    - Сортировка по имени, по рейтингу (3 ч) - около 2 часов фактически

- Экран Профиля  (8 ч)
    - Верстка (4 ч)
    - WebView сайта пользователя (2 ч)
    - Коллекция NFT (2 ч)

- Экран Коллекции (7 ч)
    - Верстка (4 ч)
    - Добавление в корзину, удаление из корзины (3 ч)
	
- Переходы по экранам (2 ч)
 Итого: 24 часа

Depechist committed:
Rating screen UI

Сверстан экран Рейтинга:

- UITableView рейтинга
- NavBarItem кнопка Сортировки
- ActionSheet по кнопке Сортировки

Depechist committed:
Rating screen UI + Network

- Небольшой рефакторинг под MVVM
- Добавлена работа с сетью
- Моковые ячейки таблицы заменены на подгружаемые данные из сети
- Добавлена логика сортировки таблицы по рейтингу и по имени (по умолчанию - по рейтингу)

Depechist committed:
statistics_epic 1/3 ready

- Небольшой рефакторинг
- Локализация строк в Action Sheet и имени экрана в TabBar

Nikitaaeee commented on FakeNFT/Scenes/Rating/RatingCell.swift:
```diff
@@ -0,0 +1,97 @@
+            make.right.equalToSuperview().offset(-16)
+            make.centerY.equalToSuperview()
+        }
+    }
```
Сейчас метод setupUI выглядит очень перегруженным, так как он выполняет очень много действий.

Рекомендую при верстке кодом использовать следующий флоу:
- Создавать элементы как переменные:
```
private let label: UILabel = {
   let label = UILabel()
   label.translatesAutoresizingMaskIntoConstraints = false
   label... (все необходимые настройки элемента)
   return label
}()
```
- Создать приватный метод _addSubViews()_ / _setupViews()_ , в котором добавить нужные элементы:
`view.addSubview(label)`
- Создать приватный метод _configureConstraints()_ / _setupConstraints())_ , в котором добавить все нужные констрейнты: 
```
NSLayoutConstraint.activate([
   label.centerYAnchor.constraint(equalTo: view.centerYAnchor)
   ...
])
```
- Вызвать эти два метода во _viewDidLoad_()

Таким образом каждый метод будет выполнять только одну задачу, а настройка UI компонентов будет происходить сразу при создании (_именно так чаще всего делают в коммерческих проектах_) 😊

Nikitaaeee commented on FakeNFT/Scenes/Rating/RatingCell.swift:
```diff
@@ -0,0 +1,97 @@
+        ratingLabel.font = UIFont.headline22
+
+        contentContainer.snp.makeConstraints { make in
+            make.top.equalToSuperview()
+            make.left.equalToSuperview().offset(35)
+            make.bottom.equalToSuperview().offset(-8)
+            make.right.equalToSuperview()
+        }
+
+        numberLabel.snp.makeConstraints { make in
+            make.right.lessThanOrEqualTo(contentContainer.snp.left).offset(-16)
+            make.centerY.equalToSuperview()
+        }
+
+        userImageView.snp.makeConstraints { make in
+            make.left.equalToSuperview().offset(16)
+            make.centerY.equalToSuperview()
+            make.height.width.equalTo(28)
+        }
+
+        nameLabel.snp.makeConstraints { make in
+            make.left.equalTo(userImageView.snp.right).offset(8)
+            make.centerY.equalToSuperview()
+        }
+
+        ratingLabel.snp.makeConstraints { make in
+            make.right.equalToSuperview().offset(-16)
+            make.centerY.equalToSuperview()
```
- Можно попробовать вынести все константы в отдельную сущность в рамках класса
Например:
```
// MARK: - Constants

private extension RatingCell {
   enum Constants {
      static let userImageViewLeadingOffset: CGFloat = 16
      ...
   }
}
```
И затем в коде вместо чисел использовать `Constants. userImageViewLeadingOffset`.
А если в коде есть какие-либо повторяющиеся числа для констрейнтов - можно создать для них отдельные св-ва, например **defaultLeadingOffset**.

- Рекомендую стараться вместо left и right при верстке использовать leading и trailing. 
Вдруг в приложении появится арабская локализация)

Nikitaaeee commented on FakeNFT/Scenes/Rating/RatingCell.swift:
```diff
@@ -0,0 +1,97 @@
+    }
+}
+
+extension UIImageView {
+    func loadImage(from url: URL) {
+        URLSession.shared.dataTask(with: url) { data, _, error in
+            if let data = data, error == nil {
+                DispatchQueue.main.async {
+                    self.image = UIImage(data: data)
+                }
+            }
+            // Обработать ошибки и плейсхолдеры для изображений (при необходимости)
+        }.resume()
+    }
+}
```
- Для расширений обычно создают отдельную папку **Extensions**. 
В ней можно создавать файлы для расширений разных типов, например файл `UIImageView+Extensions`, в котором можно хранить все расширения на **UIImageView**

- Сейчас UI компонент **UIImageView** наделен "властью" использовать метод `loadImage(from:)` и самостоятельно залезать в _URLSession_ для загрузки фото, это может создавать нежелательные зависимости и затруднять тестирование кода. 

Лучше использовать менеджер изображений или библиотеку для асинхронной загрузки, такую как **Kingfisher** или **SDWebImage**, которые обрабатывают кэширование и отмену запросов изображений.

Nikitaaeee commented on FakeNFT/Scenes/Rating/RatingViewModel.swift:
```diff
@@ -0,0 +1,81 @@
+//  Created by Artem Adiev on 05.11.2023.
+//
+
+import Foundation
+
+class RatingViewModel {
+    // MARK: - Properties
```
- В коде ViewModel нет import UIKit - круто! Значит, вся UI логика осталась в контроллере, все верно 🙌

- Рекомендую "закрывать" вью модель протоколом - создавать протокол, в котором будут все методы и свойства, к котором ты планируешь обращаться из контроллера. 
А в контроллере создавать вью модель с типом протокола:
`viewModel: RatingViewModelProtocol = RatingViewModel()`
Грамотное использование протоколов - очень важная вещь, на которую могут обращать внимание потенциальные работодатели при просмотра проектов кандидата на GitHub.

- Для RatingViewModel можно добавить **final**.
В таком случае комплятор будет быстрее находить адрес ячейки памяти при вызове функции.

Если интересно, тут можно подробнее почитать (или освежить знания) про то, что такое **диспетчеризация** и как она работает (довольно частая тема на собеседованиях): 
https://habr.com/ru/companies/simbirsoft/articles/673636/

Nikitaaeee commented on FakeNFT/Scenes/Rating/RatingViewModel.swift:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    // MARK: - Data Fetching
+    // Получаем пользователей из сети
+    func fetchUsers() {
+        guard let url = UsersRequest().endpoint else { return }
+
+        URLSession.shared.dataTask(with: url) { data, _, error in
+            if let error = error {
+                print("Error fetching users: \(error.localizedDescription)")
+                return
+            }
+            // Проверка наличия данных
+            guard let data = data else {
+                print("No data received.")
+                return
+            }
+            // Вытаскиваем данные
+            do {
+                // Декодирование полученных данных в массив структур Users
+                let decodedUsers = try JSONDecoder().decode([Users].self, from: data)
+                // Сохраняем декодированные данные в свойство users
+                self.users = decodedUsers
+                // Сортируем пользователей по рейтингу
+                self.sortByRating()
+                // Обновляем UI в главном потоке
+                DispatchQueue.main.async {
+                    self.reloadTableViewClosure?()
+                }
+            } catch {
+                // Обрабатываем ошибки
+                print("Error decoding users: \(error.localizedDescription)")
+            }
+        }.resume()
+    }
```
- Комментарии могут быть очень полезными, и я заметил, что ты прилагаешь усилия к тому, чтобы добавлять пояснения к некоторым методам/свойствам.
Однако, я хотел бы предложить более умеренный подход к комментированию. Хорошо структурированный и читаемый код часто говорит сам за себя. Если название метода или свойства ясно отражает его назначение и его поведение легко понять из контекста, то комментарий может быть не нужен. (не считая комментариев `// MARK: - ...`, их все используют 😀)
Эта тема хорошо разобрана в книге "Чистый код" Роберта Мартина.
Также про комментарии в коде можно посмотреть [тут](https://habr.com/ru/companies/enterra/articles/182574/).

- Сейчас метод `fetchUsers()` напрямую обращается к **URLSession** из Вью Модели.
Предлагаю рассмотреть возможность разделения ответственности и внедрения абстракции для сетевых запросов.
Лучше, чтобы сам запрос хранился как Дженерик в отдельном сервисе (например, _NetworkService_), тогда вью модель будет обращаться к этому сервису и вызывать его метод, а через completion возвращать результат или ошибку.

Nikitaaeee commented on FakeNFT/Scenes/Rating/RatingViewController.swift:
```diff
@@ -0,0 +1,112 @@
+import SnapKit
+
+final class RatingViewController: UIViewController {
+    private var viewModel = RatingViewModel()
+
+    // MARK: UI elements
+    private var sortButton: UIBarButtonItem?
+
+    private lazy var tableView: UITableView = {
+        let tableView = UITableView()
+        tableView.separatorStyle = .singleLine
+        tableView.layer.cornerRadius = 16
+        return tableView
+    }()
+
+    // MARK: Lifecycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        viewModel = RatingViewModel()
```
Вью модель инициализируются дважды (12я строка и 28я), это может привести к ошибкам 👀

Nikitaaeee commented on FakeNFT/Scenes/Rating/RatingViewModel.swift:
```diff
@@ -0,0 +1,81 @@
+                self.users = decodedUsers
+                // Сортируем пользователей по рейтингу
+                self.sortByRating()
+                // Обновляем UI в главном потоке
+                DispatchQueue.main.async {
+                    self.reloadTableViewClosure?()
+                }
```
Лучше все действия с **UI** производить в контроллере, в том числе и `DispatchQueue.main.async { }`

Nikitaaeee commented on FakeNFT/Scenes/Rating/RatingViewModel.swift:
```diff
@@ -0,0 +1,81 @@
+                    self.reloadTableViewClosure?()
+                }
+            } catch {
+                // Обрабатываем ошибки
+                print("Error decoding users: \(error.localizedDescription)")
+            }
```
По базовым критериям нужно будет обработать ошибки в сетевых запросах, было бы круто увидеть это в следующих итерациях 😊

Depechist committed:
Refactoring

Рефакторинг согласно замечаниям на ревью:

- Исправлен радиус закругления у аватарок пользователей
- Установка и настройка UI в RatingCell разбита на отдельные методы
- Изменен метод установки аватарки пользователя в таблице. Вместо extension UIImageView использована загрузка картинки через Kingfisher напрямую при установке изображения в методе config()
- Class RatingViewModel > Final class RatingViewModel
- Вместо создания URLSession в ViewModel использован существующий метод NetworkClient.send
- Исправлена ошибка с задублированным объявлением ViewModel в Контроллере

Depechist committed:
ProgressHUD

- Согласно замечанию на ревью на время загрузки данных Рейтинга добавлен ProgressHUD
